src/v/compat/run.cc

```diff
@@ -0,0 +1,173 @@
+            w.EndObject();
+        }
+        w.EndArray();
+        w.EndObject();
+    }
+
+    /*
+     * Writes all test cases to the provided directory.
+     */
+    static ss::future<> write(const std::filesystem::path& dir) {
+        size_t instance = 0;
+
+        for (auto& test : checker::create_test_cases()) {
+            // json encoded test case
+            auto buf = json::StringBuffer{};
+            auto writer = json::Writer<json::StringBuffer>{buf};
+            write_test_case(std::move(test), writer);
+
+            // save test case to file
+            iobuf data;
+            data.append(buf.GetString(), buf.GetSize());
+            auto fn = fmt::format("{}_{}.json", checker::name, instance++);
+            write_fully(dir / fn, std::move(data)).get();
+        }
+
+        return ss::now();
+    }
+
+    /*
+     * Check a test case.
+     */
+    static void check(json::Document doc) {
+        // fields is the content of the test instance
+        vassert(doc.HasMember("fields"), "document doesn't contain fields");
+        vassert(doc["fields"].IsObject(), "fields is not an object");
+        auto instance = checker::from_json(doc["fields"].GetObject());
+
+        // binary encodings of the test instance
+        vassert(doc.HasMember("binaries"), "document doesn't contain binaries");
+        vassert(doc["binaries"].IsArray(), "binaries is not an array");
+        auto binaries = doc["binaries"].GetArray();
+
+        for (const auto& encoding : binaries) {
+            vassert(encoding.IsObject(), "binaries entry is not an object");
+            auto binary = encoding.GetObject();
+
+            vassert(binary.HasMember("name"), "encoding doesn't have name");
+            vassert(binary["name"].IsString(), "encoding name is not string");
+            vassert(binary.HasMember("data"), "encoding doesn't have data");
+            vassert(binary["data"].IsString(), "encoding data is not string");
+
+            compat_binary data(
+              binary["name"].GetString(),
+              bytes_to_iobuf(base64_to_bytes(binary["data"].GetString())));
+
+            // keep copy for next round
+            auto&& [ia, ib] = compat_copy(std::move(instance));
+            instance = std::move(ia);
+
+            if (!checker::check(std::move(ib), std::move(data))) {
+                json::StringBuffer buf;
+                json::PrettyWriter<json::StringBuffer> writer(buf);
+                doc.Accept(writer);
+                fmt::print("Input JSON: {}\n", buf.GetString());
+                throw std::runtime_error(
+                  fmt::format("Failed checking {{{}}}\n", checker::name));
+            }
+        }
+    }
+};
+
+static void check(json::Document doc) {
+    vassert(doc.HasMember("name"), "doc doesn't have name");
+    vassert(doc["name"].IsString(), "name is doc is not a string");
+    auto name = doc["name"].GetString();
+    (void)name;
+
+    /*
+     * Add for each corpus_helper<...> type
```

Comment:
Maybe `::check` and the source in `write_corpus` can get their types from a meta typelist, iterating over them calling the respective methods on each one. It would avoid programmer error from writing some corpus code and forgetting to manually add the correct source to both of these methods.

Reply:
yeh, let me give a shot. i tried once before but it started to turn into a time sink

Reply:
+1, this will be super helpful..

Reply:
done